packages/eslint-plugin/src/rules/consistent-generic-constructors.ts

```diff
@@ -0,0 +1,84 @@
+import { AST_NODE_TYPES, TSESTree } from '@typescript-eslint/utils';
+import { createRule } from '../util';
+
+type MessageIds = 'preferLHS' | 'preferRHS';
+type Options = ['lhs' | 'rhs'];
+
+export default createRule<Options, MessageIds>({
+  name: 'consistent-generic-constructors',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Enforce specifying generic type arguments on LHS or RHS of constructor call',
+      recommended: false,
+    },
+    messages: {
+      preferLHS:
+        'The generic type arguments should be specified on the left-hand side of the constructor call.',
+      preferRHS:
+        'The generic type arguments should be specified on the right-hand side of the constructor call.',
+    },
+    fixable: 'code',
+    schema: [
+      {
+        enum: ['lhs', 'rhs'],
+      },
+    ],
+  },
+  defaultOptions: ['rhs'],
+  create(context, [mode]) {
+    return {
+      VariableDeclarator(node: TSESTree.VariableDeclarator): void {
```

Comment:
Yeah the types for these queries can be tricky. More support of us using fancy template literal magicks to parse them (#4065).

Going off the "complex types are from complex code" philosophy: how about splitting this into two queries? `VariableDeclarator[init.callee.type='Identifier'][init.type='NewExpression']` could run just the right-hand-side logic.

Search `NodeWithModifiers` and `TypeParameterWithConstraint` for examples of places where we've had to manually adjust node types in the past.

Reply:
I'm not particularly good at AST selectors, so I always felt a bit cornered when I have to use them...

Reply:
I tried, it was ugly refined types like

```ts
type TargetVariableDeclarator = TSESTree.VariableDeclarator & {
  init: TSESTree.VariableDeclarator["init"] & {
    type: AST_NODE_TYPES.NewExpression;
    callee: Extract<TSESTree.VariableDeclarator["init"], {callee: unknown}>['callee'] & {
      type: AST_NODE_TYPES.Identifier;
    }
  };
};
```

Combined with ugly selectors😇 I suggest we don't do it?